File: leetcode/242.ValidAnagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
const isAnagram = function (s, t) {
  if (s.length !== t.length) return false

  const sMap = s.split('').reduce((acc, str) => {
    acc[str] = 1 + (acc[str] || 0)
    return acc
  }, {})

  const tMap = t.split('').reduce((acc, str) => {
    acc[str] = 1 + (acc[str] || 0)
    return acc
  }, {})

  for (const key in sMap) {
    if (sMap[key] !== tMap[key]) return false
  }

  return true
}
console.log(isAnagram('rat', 'car'))
console.log(isAnagram('anagram', 'nagaram'))

File: leetcode/242.ValidAnagram.kt
class Solution {
	fun isAnagram(s: String, t: String): Boolean {
		if (s.length != t.length) return false

		val sMap = s.groupingBy { it }.eachCount()
		val tMap = t.groupingBy { it }.eachCount()

		for (char in t) {
			if (sMap[char] != tMap[char]) return false
		}

		return true
	}
}
